src/Build/Instance/ProjectInstance.cs

```diff
@@ -494,6 +494,38 @@ private ProjectInstance(Project linkedProject, bool fastItemLookupNeeded)
             _explicitToolsVersionSpecified = linkedProject.SubToolsetVersion != null;
 
             _isImmutable = true;
+
+            static List<string> GetImportFullPaths(ObjectModelRemoting.ProjectLink projectLink)
+            {
+                // Imports collection contains ResolvedImports, which are structures which are much bigger than the string list.
+                // We convert to the string list and let Imports itself to BE GCed.
+                var imports = projectLink.Imports;
+                var paths = new List<string>(imports.Count);
+                foreach (var import in imports)
+                {
+                    if (import.ImportedProject != null)
```

Comment:
Is import.ImportedProject almost never null? If not, would it be worth it to size the paths list based on projectLink.Imports.Count(import => import.ImportedProject != null)?